### PR TITLE
Improve SelectBase coverage

### PR DESCRIPTION
**Pull request by timowest**
Fixes https://github.com/querydsl/querydsl/issues/1204

Databases to go through

* [x] CUBRID 9.2.0
* [x] DB2 10.1.2
* [x] Derby 10.11.1.1
* [x] Firebird 2.5.2
* [x] H2 1.4.178
* [x] HSQLDB 2.3.2
* [x] MSSQL 2008
* [x] MySQL 5.5
* [x] Oracle 11.2.0
* [x] PostgreSQL 9.3
* [X] SQLite
* [x] Teradata 14.0.3

DB2, MSSQL, Oracle and Teradata tests are not run in travis

backport https://github.com/querydsl/querydsl/pull/1269

### CODE REVIEW

**Comment by timowest** [comment]
Databases to go through

* [ ] CUBRID
* [ ] DB2
* [ ] Derby
* [ ] Firebird
* [ ] H2
* [ ] HSQLDB
* [ ] MSSQL
* [ ] MySQL
* [ ] Oracle
* [ ] PostgreSQL
* [ ] SQLite
* [ ] Teradata

**Comment by timowest** [comment]
I rebased this now.

**Comment by timowest** [comment]
@johnktims @Shredder121 This is ready to review now.

**Comment by timowest** [comment]
@johnktims @Shredder121 How does this look?

File: pom.xml [hunk 1]
```diff
@@ -27,14 +27,14 @@
     <project.checkout>scm:git:git@github.com:querydsl/querydsl.git</project.checkout>
     
     <!-- deps -->
-    <derby.version>10.10.1.1</derby.version>
-    <hsqldb.version>2.3.1</hsqldb.version>
+    <derby.version>10.11.1.1</derby.version>
+    <hsqldb.version>2.3.2</hsqldb.version>
     <h2.version>1.4.178</h2.version>
```
Comment (johnktims) [inline pom.xml]: 1.4.186 is the latest version for H2.

There are some other dependencies that could be updated. Separate PR, perhaps?

File: querydsl-sql-spatial/src/main/java/com/querydsl/sql/spatial/SQLServerGeometryType.java [hunk 2]
```diff
@@ -72,7 +74,7 @@ public String getLiteral(Geometry geometry) {
         if (geometry.getSRID() > -1) {
             return "geometry::STGeomFromText('" + str + "', " + geometry.getSRID() + ")";
         } else {
-            return "geometry::STGeomFromText('" + str + "')";
+            return "geometry::STGeomFromText('" + str + "', 4326)";
```
Comment (johnktims) [inline querydsl-sql-spatial/src/main/java/com/querydsl/sql/spatial/SQLServerGeometryType.java]: What is "4326"?

Comment (Shredder121) [inline querydsl-sql-spatial/src/main/java/com/querydsl/sql/spatial/SQLServerGeometryType.java]: SQLServer's default `geometry.getSRID()`?

File: pom.xml (hunk 1 above)
Comment (timowest) [inline pom.xml]: I can update it here as well, but for others a separate PR could be created.

File: querydsl-sql-spatial/src/main/java/com/querydsl/sql/spatial/SQLServerGeometryType.java (hunk 2 above)
Comment (timowest) [inline querydsl-sql-spatial/src/main/java/com/querydsl/sql/spatial/SQLServerGeometryType.java]: The SRID 4326 stands for WGS 84 http://en.wikipedia.org/wiki/World_Geodetic_System#A_new_World_Geodetic_System:_WGS_84

It is used as the default SRID elsewhere https://msdn.microsoft.com/en-us/library/bb933824.aspx

File: querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java [hunk 3]
```diff
@@ -45,7 +45,29 @@
 
     @Override
     public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-        if (method.getName().startsWith("get")) {
+        if (method.getName().equals("wasNull")) {
```
Comment (johnktims) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: It doesn't matter here since it's just a test, but I like to call `equals` on the constant string (`"wasNull".equals(method.getName());`) to eliminate any chance of an NPE.

File: querydsl-sql-spatial/src/main/java/com/querydsl/sql/spatial/SQLServerGeometryType.java (hunk 2 above)
Comment (johnktims) [inline querydsl-sql-spatial/src/main/java/com/querydsl/sql/spatial/SQLServerGeometryType.java]: I would probably name that magic constant, but I think the rest of the PR looks good.

File: querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java (hunk 3 above)
Comment (timowest) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: How does that protect of an NPE in this case? `method.getName()` is guaranteed to be non null, and the reverse order doesn't handle the `method == null` case.

Comment (Shredder121) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: I think he means it more as a [general rule](http://checkstyle.sourceforge.net/config_coding.html#EqualsAvoidNull).
Since there is a checkstyle check, we could incorporate this.

File: querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java [hunk 4]
```diff
@@ -395,7 +395,7 @@ public void Custom_Projection() {
     }
 
     @Test
-    @IncludeIn({H2, SQLSERVER, MYSQL, ORACLE, TERADATA}) // TODO fix postgresql
+    @ExcludeIn({CUBRID, DB2, DERBY, HSQLDB, POSTGRESQL, SQLITE, TERADATA})
```
Comment (Shredder121) [inline querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java]: I suggest that we also apply this on the other branch, so we don't pull our hairs out every time we change this line.
We could also make it a separate PR to equalize this everywhere.

File: querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java (hunk 3 above)
Comment (timowest) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: I'd consider the inverted order only for known nullable types and not as a general practice, since having the non-null visible in the access is also valuable and at least for me this order is more readable.

File: querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java (hunk 4 above)
Comment (timowest) [inline querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java]: what do you mean? The ExcludeIn usage for this method is the same on the other branch.

Comment (Shredder121) [inline querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java]: Also in that order? If that is the case, then it's okay.

Comment (timowest) [inline querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java]: it's POSTGRES in querydsl3 and POSTGRESQL in querydsl4, otherwise the same

File: querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java (hunk 3 above)
Comment (Shredder121) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: Personally I find the order as it is now much more readable.
If I want to do a null-safe equals, I always use Guava's `Objects.equal` instead.

Comment (timowest) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: good point with `Objects.equals`, which solves the more general case

Comment (Shredder121) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: *__Not__ `Objects.equals`, that is `java.util.Objects` and is [not available in java 6](https://github.com/querydsl/querydsl/issues/1162)*

Comment (timowest) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: fixed

File: querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java (hunk 4 above)
Comment (Shredder121) [inline querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java]: Still impossible to cherry pick, but it's okay.
Maybe it's a good idea to always order it alphabetically? (I don't know if there is a checkstyle check for that?)

File: querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java (hunk 3 above)
Comment (johnktims) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: I wasn't aware of `Objects.equal`. I should have known guava would have a helper method for this!

Comment (Shredder121) [inline querydsl-sql/src/test/java/com/querydsl/sql/types/TypeTest.java]: Yes, they have helper methods for a lot of things, it's knowing (and using) them that makes Guava so powerful.
However, the initial discoveries can be quite slow, since a lot of helper methods are very domain specific (`io` package is pretty common, but I don't use `net` for example)

File: querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java (hunk 4 above)
Comment (timowest) [inline querydsl-sql/src/test/java/com/querydsl/sql/SelectBase.java]: Yes, we can maybe homogenize that later.

**Comment by timowest** [comment]
@johnktims @Shredder121 Is this good to merge, once the tests pass?

**Comment by johnktims** [comment]
:+1:

**Comment by Shredder121** [comment]
Yes, for me it looks good as well. :+1: 

Just a question, is there a reason we explicitly box primitives?

**Comment by timowest** [comment]
Not a good reason for boxing, but that could be maybe removed in another PR?

**Comment by Shredder121** [comment]
I didn't really mean so we could 'fix' that,  more if there was something behind it.
